File: 1-easy/00020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for ch in s:
            match ch:
                case '}' | ']' | ')':
                    if len(stack) == 0:
                        return False

                    top_char = stack[-1]

                    if (ch == '}' and top_char == '{') \
                        or (ch == ']' and top_char == '[') \
                        or (ch == ')' and top_char == '('):
                        stack.pop()
                    else:
                        return False

                case _:
                    stack.append(ch)

        return len(stack) == 0

File: 1-easy/00020-valid-parentheses/solution.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
    
        let mut stack: Vec<u8> = Vec::new();

        for byte in s.into_bytes().into_iter() {
            match byte {
                b'}' | b']' | b')' => {
                    if let Some(top_byte) = stack.pop() {
                        if (byte == b'}' && top_byte == b'{')
                            || (byte == b']' && top_byte == b'[')
                            || (byte == b')' && top_byte == b'(')
                        {
                            //Do nothing
                        } else {
                            return false;
                        }
                    } else {
                        return false;
                    }
                },
                _ => {
                    stack.push(byte);
                },
            } 
        }

        stack.len() == 0
    }
}
